libraries/chain/controller.cpp

```diff
@@ -3473,14 +3473,13 @@ struct controller_impl {
    void consider_voting(const block_state_ptr& bsp) {
       // 1. Get the `core.final_on_strong_qc_block_num` for the block you are considering to vote on and use that to find the actual block ID
       //    of the ancestor block that has that block number.
-      // 2. If that block ID is not an ancestor of the current head block, then do not vote for that block.
+      // 2. If that block ID is for a non validated block, then do not vote for that block.
       // 3. Otherwise, consider voting for that block according to the decide_vote rules.
 
       if (bsp->core.final_on_strong_qc_block_num > 0) {
```

Comment:
```suggestion
      if (bsp->is_valid() && bsp->core.final_on_strong_qc_block_num > 0) {
```

Reply:
No, don't need the block we vote on to be valid. It will not be valid at this point because we just received the block. The point of this is to vote early before we validate the block.

Reply:
I can't remember the reasoning as to why we don't need a block to be validated to vote on it? Don't we run the risk that is a bad block is gossipped, many finalizers will vote on it and it will get a QC even though it doesn't validate?

Reply:
We don't vote to make something final unless it has been validated. So ultimately we do not contribute to finality on a block unless we have validated it. The voting on non-validated blocks just means we have verified the block header and we anticipate it being a valid block.

Reply:
> we anticipate it being a valid block.

which means we are not sure, right? So we would have voted to make an unvalidated block final.

Reply:
We have not voted to make an unvalidated block final. We have voted to lock on a block that has not been validated. We  can switch to voting on a diff branch later. Although, I think the test failures on this PR may indicate that switching to a different branch to vote later is not working.

Reply:
Thanks for explaining. I finally get the intent of this PR. I am not convinced that it is a wise change to make (have we proven that it would work in all cases?). If no more QCs are produced I don't think we'll regain liveness or safety.